File: apps/models/news.py
# -*- encoding: utf-8 -*-
"""
Copernicus Operations Dashboard

Copyright (C) ${startYear}-${currentYear} ${Telespazio}
All rights reserved.

This document discloses subject matter in which TPZ has
proprietary rights. Recipient of the document shall not duplicate, use or
disclose in whole or in part, information contained herein except for or on
behalf of TPZ to fulfill the purpose for which the document was
delivered to him.
"""

import logging
from datetime import datetime

from apps import db
from apps.utils.db_utils import generate_uuid

logger = logging.getLogger(__name__)


class News(db.Model):
    __tablename__ = 'news'

    id = db.Column(db.String(64), primary_key=True)
    title = db.Column(db.String(9999))
    text = db.Column(db.String(9999))
    link = db.Column(db.String(9999))
    publicationDate = db.Column(db.DateTime)
    occurrenceDate = db.Column(db.DateTime)
    category = db.Column(db.String(999))
    impactedSatellite = db.Column(db.String(99))
    environment = db.Column(db.String(9999))
    datatakes_completeness = db.Column(db.String(9999))
    modifyDate = db.Column(db.DateTime)

    def __init__(self, **kwargs):
        for property, value in kwargs.items():
            if hasattr(value, '__iter__') and not isinstance(value, str):
                value = value[0]

            setattr(self, property, value)


def save_news(title, text, link, publication_date, occurrence_date, category, impacted_satellite, environment,
              datatakes_completeness, modify_date=datetime.now()):
    try:
        news = News(id=str(generate_uuid()), title=title, text=text, link=link, publicationDate=publication_date,
                    occurrenceDate=occurrence_date, category=category, impactedSatellite=impacted_satellite,
                    environment=environment, datatakes_completeness=str(datatakes_completeness), modifyDate=modify_date)
        db.session.add(news)
        db.session.commit()
        return news
    except Exception as ex:
        db.session.rollback()
    return None


def update_news(title, text, link, publication_date, occurrence_date, category, impacted_satellite, environment,
              modify_date=datetime.now()):
    try:
        news = db.session.query(News).filter(News.link == link).first()
        if news is not None:
            news.title = title
            news.text = text
            news.publicationDate = publication_date
            news.modifyDate = modify_date
        else:
            datatakes_completeness = []
            if environment is not None and len(environment) > 0:
                datatake_ids = environment.split(';')
                for datatake_id in datatake_ids:
                    if datatake_id is None or len(datatake_id) == 0:
                        continue
                    entry = {'datatakeID': datatake_id, 'L0_': 0, 'L1_': 0, 'L2_': 0}
                    datatakes_completeness.append(entry)
            news = News(id=str(generate_uuid()), title=title, text=text, link=link, publicationDate=publication_date,
                        occurrenceDate=occurrence_date, category=category, impactedSatellite=impacted_satellite,
                        environment=environment, datatakes_completeness=str(datatakes_completeness),
                        modifyDate=modify_date)
            db.session.add(news)
        db.session.commit()
        return news
    except Exception as ex:
        db.session.rollback()
    return None


def update_news_categorization(link, category, impacted_satellite, environment, occurrenceDate):
    try:
        news = db.session.query(News).filter(News.link == link).first()
        if news is not None:
            news.category = category
            news.impactedSatellite = impacted_satellite
            news.environment = environment
            news.occurrenceDate = datetime.strptime(occurrenceDate, '%d/%m/%Y %H:%M:%S')
            db.session.commit()
            return news
    except Exception as ex:
        db.session.rollback()
    return None


def update_datatakes_completeness(link, datatakes_completeness):
    try:
        news = db.session.query(News).filter(News.link == link).first()
        if news is not None:
            news.datatakes_completeness = str(datatakes_completeness)
            news.modify_date = datetime.now()
            db.session.add(news)
            db.session.commit()
            return news
        else:
            return None
    except Exception as ex:
        db.session.rollback()
        return None


def get_news(start_date=None, end_date=None):
    try:
        if start_date is None or end_date is None:
            return News.query.order_by(News.occurrenceDate.asc()).all()
        else:
            return News.query.filter(News.publicationDate is not None). \
                filter(News.publicationDate >= start_date).filter(News.publicationDate <= end_date).order_by(
                News.occurrenceDate.asc()).all()
    except Exception as ex:
        logger.error("Retrieving News, received error: %s", ex, exc_info=True)
        return None


def get_news_by_information(category, impacted_satellite, occurrence_date):
    try:
        return News.query.filter_by(category=category, impactedSatellite=impacted_satellite,
                                    occurrenceDate=occurrence_date).order_by(News.modifyDate.asc()).all()
    except Exception as ex:
        return None


def get_news_by_environment(environment):
    try:
        search = "%{}%".format(environment)
        return db.session.query(News).filter(News.environment.like(search)).order_by(
            News.modifyDate.asc()).all()
    except Exception as ex:
        return None


def delete_news_by_id(uuid):
    try:
        db.session.query(
            News
        ).filter(
            News.id == uuid,
        ).delete()

        db.session.commit()
    except Exception as ex:
        pass
    return

File: apps/static/assets/js/admin/news.js
/*
Copernicus Operations Dashboard

Copyright (C) ${startYear}-${currentYear} ${Telespazio}
All rights reserved.

This document discloses subject matter in which TPZ has
proprietary rights. Recipient of the document shall not duplicate, use or
disclose in whole or in part, information contained herein except for or on
behalf of TPZ to fulfill the purpose for which the document was
delivered to him.
*/

class News {

    // Move date handling to MIXIN (periodSelection)
    constructor() {

        // Hide time period selector
        $('#time-period-select').hide();

        // News table
        try {
            this.newsTable = $('#basic-datatables-news').DataTable({
                "language": {
                  "emptyTable": "Retrieving news..."
                },
                columnDefs: [{
                        targets: -1,
                        data: null,
                        render: function (data, type, row) {
                            if (type === 'display') {
                                return '<button type="button" class="btn-link" data-toggle="modal" data-target="#editNewsModal" '+
                                    'onclick="news.editNewsDetails(\'' + data[4] + '\')"><i class="icon-pencil"></i></button>';
                            } else {
                                return data;
                            }
                        }
                    }]
                });
        } catch(err) {
            console.info('Initializing news class - skipping table creation...')
        }

        // Array containing serialized anomaly categories
        this.categories = [
            'Acquisition',
            'Platform',
            'Production',
            'Manoeuvre',
            'Calibration',
            'Data access',
            'Archive'];

        // Array containing serialized impacted satellites
        this.impactedSatellites = [
            'Copernicus Sentinel-1A',
            'Copernicus Sentinel-2A',
            'Copernicus Sentinel-2B',
            'Copernicus Sentinel-3A',
            'Copernicus Sentinel-3B',
            'Copernicus Sentinel-5p',];

        // Map containing serialized anomalies accessed from "key" field
        this.newsList = {};
    }

    init() {

        // Retrieve the anomalies from local MYSQL DB
        asyncAjaxCall('/api/events/news/previous-quarter', 'GET', {}, news.successLoadNews.bind(this),
            news.errorLoadNews.bind(this));

        return;
    }

    successLoadNews(response) {

        // Acknowledge the successful retrieval of news
        var rows = format_response(response);
        console.info('News successfully retrieved');
        console.info("Number of records: " + rows.length);

        // Parse response
        var data = new Array();
        for (var i = 0 ; i < rows.length ; ++i){

            // Auxiliary variables
            // Use the link as the unique key identifier (human readable)
            var newsItem = rows[i];
            var key = newsItem['link'];

            // Save a local copy of the news
            news.newsList[key] = newsItem;

            // Push the element row, with the collected information
            // row is a datatable row, related to a single user
            // News status record:
            // Title, occurrenceDate, category, impactedSatellite, link
            data.push([newsItem['title'], newsItem['occurrenceDate'], newsItem['category'],
                newsItem['impactedSatellite'], newsItem['link']]);
        }

        // Refresh news table and return
        news.newsTable.clear().rows.add(data).draw();
        return;
    }

    errorLoadNews(response){
        console.error(response)
        return;
    }

    editNewsDetails(key) {
        let newsItem = news.newsList[key];
        news.buildNewsDetailsPanel(newsItem);
    }

    buildNewsDetailsPanel(newsItem) {

        // Build widgets
        // Title
        $('#news-details').html('');
        $('#news-details').append(
            '<div class="form-group" id="news-title-div" style="width: 430px">' +
                '<label for="news-title">News title</label>' +
                '<input type="text" disabled style="color: #6c757d" class="form-control" id="news-title" placeholder="news title">' +
            '</div>');
        $('#news-title').val(newsItem['title']);

        // Link
        $('#news-details').append(
            '<div class="form-group" id="news-link-div" style="width: 430px">' +
                '<label for="news-link">Link</label>' +
                '<input type="text" disabled style="color: #6c757d" class="form-control" id="news-link" placeholder="news link">' +
            '</div>');
        $('#news-link').val(newsItem['link']);

        // Category select
        $('#news-details').append(
            '<div class="form-group">' +
                '<label for="news-category-select">Category</label>' +
                '<select class="form-control" id="news-category-select" placeholder="category"></select>' +
            '</div>');
        let selectedCategory = null
        news.categories.forEach(function(category) {
            if (newsItem['category'] === category) selectedCategory = category;
            let selected = newsItem['category'] === category ? ' selected ' : '';
            $('#news-category-select').append('<option value="' + category + '"' + selected + '>' + category + '</option>');
        });

        // Impacted satellite select
        $('#news-details').append(
            '<div class="form-group">' +
                '<label for="news-impacted-satellite-select">Impacted Satellite</label>' +
                '<select class="form-control" id="news-impacted-satellite-select" placeholder="impacted satellite"></select>' +
            '</div>');
        news.impactedSatellites.forEach(function(satellite) {
            let selected = newsItem['impactedSatellite'] === satellite ? ' selected ' : '';
            $('#news-impacted-satellite-select').append('<option value="' + satellite + '"' + selected + '>' + satellite + '</option>');
        });

        // Occurrence date
        $('#news-details').append(
            '<div class="form-group" id="news-occurrence-date-div" style="width: 430px">' +
                '<label for="news-occurrence-date">Occurrence date</label>' +
                '<input type="text" class="form-control" id="news-occurrence-date" placeholder="dd/MM/yyyy HH:mm:ss.SSSSSS">' +
            '</div>');
        $('#news-occurrence-date').val(newsItem['occurrenceDate']);

        // Data takes
        $('#news-details').append(
            '<div class="form-group" id="news-environment-div" style="width: 430px">' +
                '<label for="news-environment">Datatakes</label>' +
                '<input type="text" class="form-control" id="news-environment" placeholder="datatakes">' +
            '</div>');
        $('#news-environment').val(newsItem['environment']);
    }

    updateNews() {

        // Retrieve modified news details
        let link = $('#news-link').val();
        let title = $('#news-title').val();
        let category = $('#news-category-select').val();
        let impactedSatellite = $('#news-impacted-satellite-select').val();
        let environment = $('#news-environment').val();
        let occurrenceDate = $('#news-occurrence-date').val();

        // Invoke user's details update
        let newsItem = news.newsList[link];
        newsItem['link'] = link;
        newsItem['title'] = title;
        newsItem['category'] = category;
        newsItem['impactedSatellite'] = impactedSatellite;
        newsItem['environment'] = environment;
        newsItem['occurrenceDate'] = occurrenceDate;
        ajaxCall('/api/events/news/update', 'POST', newsItem, this.successUpdateNews.bind(this), this.errorUpdateNews.bind(this));
    }

    successUpdateNews() {

        // Close edit anomaly modal window
        $('#editNewsModal').modal('hide');

        // Clean news details panel
        $('#news-link').val('');
        $('#news-category-select').val('');
        $('#news-impacted-satellite-select').val('');
        $('#news-link').val('');
        $('#news-environment').val('');
        $('#news-occurrence-date').val('');

        // Refresh anomalies table
        asyncAjaxCall('/api/events/news/previous-quarter', 'GET', {}, news.successLoadNews, news.errorLoadNews);
    }

    errorUpdateNews() {
        console.error(response)
        return;
    }
}

let news = new News();